File: app/models/Property.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2014 NAVER Corp.
 * http://yobi.io
 *
 * @Author Yi EungJun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package models;

import play.data.validation.Constraints;
import play.db.ebean.Model;
import utils.Diagnostic;

import javax.annotation.Nonnull;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.Id;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@Entity
public class Property extends Model {
    public static final Finder<Long, Property> find = new Finder<>(Long.class, Property.class);

    private static final long serialVersionUID = 8074682539173273921L;

    @Id
    public Long id;

    @Enumerated(EnumType.STRING)
    @Constraints.MaxLength(255)
    public Name name;

    @Constraints.MaxLength(4000)
    public String value;

    public static String get(Name name) {
        List<Property> properties = find.where().eq("name", name).findList();

        if (properties.size() > 0) {
            return properties.get(0).value;
        } else {
            return null;
        }
    }

    public static Long getLong(Name name) {
        String value = get(name);

        return value == null ? null : Long.valueOf(value);
    }

    public static void set(Name name, String value) {
        Property property = find.where().eq("name", name).findUnique();

        if (property == null) {
            property = new Property();
            property.name = name;
        }

        property.value = value;
        property.save();
    }

    public static void set(Name name, Long value) {
        set(name, value.toString());
    }

    public static enum Name {
        // the uid of the most recent email Mailbox has received
        MAILBOX_LAST_SEEN_UID,
        // the uidvalidity of the imap folder Mailbox has used most recently
        MAILBOX_LAST_UID_VALIDITY
        // Add property you need here.
    }

    public static void onStart() {
        Diagnostic.register(new Diagnostic() {
            @Override
            public @Nonnull List<String> check() {
                List<String> errors = new ArrayList<>();

                for (Property.Name name : Property.Name.values()) {
                    Set<Property> properties = Property.find.where().eq("name", name).findSet();
                    if (properties.size() > 1) {
                        errors.add(String.format("Property '%s' has duplicated values: %s",
                                name, properties));
                    }
                }

                return errors;
            }
        });
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/support/Property.kt
package com.github.yonaprojects.yona.domain.support

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint

// yona models/Property.java 대응 (P1-55). 이름-값 형태의 사이트 전역 설정을 저장하는 범용 엔티티지만,
// 실제로는 IMAP 메일함 폴링 워터마크(MAILBOX_LAST_SEEN_UID/MAILBOX_LAST_UID_VALIDITY) 용도로만
// 쓰인다 — yona 원본도 이 두 값 외에는 쓰지 않는다("Add property you need here"라는 주석만 있음).
enum class PropertyName {
    MAILBOX_LAST_SEEN_UID,
    MAILBOX_LAST_UID_VALIDITY
}

@Entity
@Table(name = "property", uniqueConstraints = [UniqueConstraint(columnNames = ["name"])])
class Property(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var name: PropertyName,

    @Column(length = 4000)
    var value: String? = null
)
